File: 17-Basic – Calculator Program/calculator-project/c_calculator/calculator.c
#include <stdio.h>
#include <stdlib.h>

// Function prototypes
double add(double a, double b);
double subtract(double a, double b);
double multiply(double a, double b);
double divide(double a, double b);
void display_menu();

int main() {
    double num1, num2, result;
    char operation;
    int choice;
    
    printf("=== BASIC CALCULATOR PROGRAM ===\n");
    
    do {
        display_menu();
        printf("Enter your choice (1-5): ");
        scanf("%d", &choice);
        
        if(choice == 5) {
            printf("Thank you for using the calculator! Goodbye!\n");
            break;
        }
        
        if(choice < 1 || choice > 5) {
            printf("Invalid choice! Please select between 1-5.\n\n");
            continue;
        }
        
        printf("Enter first number: ");
        scanf("%lf", &num1);
        printf("Enter second number: ");
        scanf("%lf", &num2);
        
        switch(choice) {
            case 1:
                result = add(num1, num2);
                printf("Result: %.2lf + %.2lf = %.2lf\n\n", num1, num2, result);
                break;
                
            case 2:
                result = subtract(num1, num2);
                printf("Result: %.2lf - %.2lf = %.2lf\n\n", num1, num2, result);
                break;
                
            case 3:
                result = multiply(num1, num2);
                printf("Result: %.2lf * %.2lf = %.2lf\n\n", num1, num2, result);
                break;
                
            case 4:
                if(num2 != 0) {
                    result = divide(num1, num2);
                    printf("Result: %.2lf / %.2lf = %.2lf\n\n", num1, num2, result);
                } else {
                    printf("Error: Division by zero is not allowed!\n\n");
                }
                break;
                
            default:
                printf("Invalid operation!\n\n");
        }
        
    } while(choice != 5);
    
    return 0;
}

// Function to display menu
void display_menu() {
    printf("Available Operations:\n");
    printf("1. Addition (+)\n");
    printf("2. Subtraction (-)\n");
    printf("3. Multiplication (*)\n");
    printf("4. Division (/)\n");
    printf("5. Exit\n");
}

// Arithmetic functions
double add(double a, double b) {
    return a + b;
}

double subtract(double a, double b) {
    return a - b;
}

double multiply(double a, double b) {
    return a * b;
}

double divide(double a, double b) {
    return a / b;
}

File: 17-Basic – Calculator Program/calculator-project/python_calculator/calculator.py
"""
Advanced Calculator Program with basic arithmetic operations
"""

class Calculator:
    """A simple calculator class implementing basic operations"""
    
    @staticmethod
    def add(a: float, b: float) -> float:
        """Add two numbers"""
        return a + b
    
    @staticmethod
    def subtract(a: float, b: float) -> float:
        """Subtract second number from first"""
        return a - b
    
    @staticmethod
    def multiply(a: float, b: float) -> float:
        """Multiply two numbers"""
        return a * b
    
    @staticmethod
    def divide(a: float, b: float) -> float:
        """Divide first number by second number"""
        if b == 0:
            raise ValueError("Division by zero is not allowed!")
        return a / b
    
    @staticmethod
    def power(a: float, b: float) -> float:
        """Raise first number to the power of second number"""
        return a ** b
    
    @staticmethod
    def modulo(a: float, b: float) -> float:
        """Calculate modulo of first number by second number"""
        if b == 0:
            raise ValueError("Modulo by zero is not allowed!")
        return a % b


def display_menu() -> None:
    """Display the calculator menu"""
    print("\n" + "="*50)
    print("           ADVANCED CALCULATOR PROGRAM")
    print("="*50)
    print("1. Addition (+)")
    print("2. Subtraction (-)")
    print("3. Multiplication (*)")
    print("4. Division (/)")
    print("5. Power (^)")
    print("6. Modulo (%)")
    print("7. Exit")
    print("="*50)


def get_number_input(prompt: str) -> float:
    """Get valid number input from user"""
    while True:
        try:
            return float(input(prompt))
        except ValueError:
            print("Invalid input! Please enter a valid number.")


def get_operation_choice() -> int:
    """Get valid operation choice from user"""
    while True:
        try:
            choice = int(input("Enter your choice (1-7): "))
            if 1 <= choice <= 7:
                return choice
            else:
                print("Please enter a number between 1 and 7.")
        except ValueError:
            print("Invalid input! Please enter a number between 1 and 7.")


def perform_operation(choice: int, calc: Calculator, num1: float, num2: float) -> None:
    """Perform the selected arithmetic operation"""
    operations = {
        1: ("+", calc.add),
        2: ("-", calc.subtract),
        3: ("*", calc.multiply),
        4: ("/", calc.divide),
        5: ("^", calc.power),
        6: ("%", calc.modulo)
    }
    
    operator, operation_func = operations[choice]
    
    try:
        result = operation_func(num1, num2)
        print(f"Result: {num1} {operator} {num2} = {result}")
    except (ValueError, ZeroDivisionError) as e:
        print(f"Error: {e}")


def main():
    """Main function to run the calculator program"""
    calc = Calculator()  # This line was missing the Calculator class definition
    
    print("Welcome to the Advanced Calculator!")
    
    while True:
        display_menu()
        choice = get_operation_choice()
        
        if choice == 7:
            print("Thank you for using the calculator! Goodbye!")
            break
        
        # Fixed the multi-line string issue
        operation_names = ['Addition', 'Subtraction', 'Multiplication', 'Division', 'Power', 'Modulo']
        print(f"\nOperation selected: {operation_names[choice-1]}")
        
        num1 = get_number_input("Enter first number: ")
        num2 = get_number_input("Enter second number: ")
        
        perform_operation(choice, calc, num1, num2)
        
        # Ask if user wants to continue
        continue_calc = input("\nDo you want to perform another calculation? (y/n): ").lower()
        if continue_calc not in ['y', 'yes']:
            print("Thank you for using the calculator! Goodbye!")
            break


if __name__ == "__main__":
    main()